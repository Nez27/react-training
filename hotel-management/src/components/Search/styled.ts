import styled from 'styled-components';

const StyledSearch = styled.input`
  border-radius: var(--radius-sm);
  border: 1px solid var(--border-color);
  font-size: var(--fs-sm-x);
  padding: 5px 10px;
`;

export { StyledSearch };
