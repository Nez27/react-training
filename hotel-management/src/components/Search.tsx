import { useEffect, useState } from 'react';
import styled from 'styled-components';

const StyledSearch = styled.input`
  border-radius: var(--radius-sm);
  border: 1px solid var(--border-color);
  font-size: var(--fs-sm-x);
  padding: 5px 10px;
`;

interface ISearch {
  setPlaceHolder: string;
  setValueSearch: React.Dispatch<React.SetStateAction<string>>;
}

const Search = ({ setValueSearch, setPlaceHolder }: ISearch) => {
  const [query, setQuery] = useState('');

  useEffect(() => {
    const timeOut = setTimeout(() => {
      setValueSearch(query);
    }, 500);

    return () => clearTimeout(timeOut);
  }, [setValueSearch, query]);

  return (
    <StyledSearch
      onChange={(e) => setQuery(e.target.value)}
      placeholder={setPlaceHolder}
    />
  );
};

export default Search;
