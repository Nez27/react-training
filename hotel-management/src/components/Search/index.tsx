import { useEffect, useState } from 'react';
import { useSearchParams } from 'react-router-dom';

// Styled
import { StyledSearch } from './styled';

// Hooks
import { useDebounce } from '@hook/useDebounce';

// Types
import { Nullable } from '@type/common';

interface ISearch {
  setPlaceHolder: string;
}

const Search = ({ setPlaceHolder }: ISearch) => {
  const field = 'search';
  const [searchParams, setSearchParams] = useSearchParams();
  const [query, setQuery] = useState<Nullable<string>>(null);
  const debounceValue = useDebounce<Nullable<string>>(query, 700);

  useEffect(() => {
    if (debounceValue !== null) {
      searchParams.set(field, debounceValue);

      setSearchParams(searchParams);
    }
  }, [debounceValue, searchParams, setSearchParams]);

  return (
    <StyledSearch
      onChange={(e) => setQuery(e.target.value)}
      placeholder={setPlaceHolder}
    />
  );
};

export default Search;
