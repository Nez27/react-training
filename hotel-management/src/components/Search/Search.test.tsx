import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';

// Components
import Search from '.';

describe('Search', () => {
  const placeHolder = 'Search placeholder...';
  const wrapper = renderer.create(
    <BrowserRouter>
      <Search setPlaceHolder={placeHolder} />
    </BrowserRouter>
  );

  test('Should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
