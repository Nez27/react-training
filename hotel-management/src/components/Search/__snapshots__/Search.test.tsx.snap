// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Should render correctly 1`] = `
<input
  className="sc-bdnyFh eAgOPb"
  onChange={[Function]}
  placeholder="Search placeholder..."
/>
`;
